presto-main/src/main/java/com/facebook/presto/operator/aggregation/AccumulatorCompiler.java

```diff
@@ -116,8 +116,7 @@ public static GenericAccumulatorFactoryBinder generateAccumulatorFactoryBinder(A
         FieldDefinition stateFactoryField = definition.declareField(a(PRIVATE, FINAL), "stateFactory", AccumulatorStateFactory.class);
         FieldDefinition inputChannelsField = definition.declareField(a(PRIVATE, FINAL), "inputChannels", type(List.class, Integer.class));
         FieldDefinition maskChannelField = definition.declareField(a(PRIVATE, FINAL), "maskChannel", type(Optional.class, Integer.class));
-        Class<?> stateClass = grouped ? stateFactory.getGroupedStateClass() : stateFactory.getSingleStateClass();
-        FieldDefinition stateField = definition.declareField(a(PRIVATE, FINAL), "state", stateClass);
+        FieldDefinition stateField = definition.declareField(a(PRIVATE, FINAL), "state", grouped ? stateFactory.getGroupedStateClass() : stateFactory.getSingleStateClass());
```

Comment:
i want to understand this, but state class already depends on grouped or not and is set to single or grouped.  how is using stateClass here any different? it's literally inlining the above code? 

the only diff I see is Class<?> variable vs types returned as Class<? extends T>

maybe that?

Reply:
@rash67 : The scratch state has very fixed usage pattern: it will first be deserialized, and then be called combine as the second parameter, see https://gist.github.com/wenleix/bc65ffe32e18d35949c52377c8c833ae#file-gistfile1-txt-L175 (I dumped from bytecode tree :) )

Thus it doesn't care whether it's a grouped state or single state, as long as it supports state interface :).

It opens opportunities to even some kind of "LazyState", basically store the (unserialized) Block and position, and do deserialize and combine in one step :)

For my case, it's simply because initialize Group State can be too expensive. A huge time is spending on deserialization (actually allocation and jlinit). And make the query 3x CPU more expensive .

![screen shot 2018-03-14 at 22 41 20](https://user-images.githubusercontent.com/799346/37446173-eb836f00-27d8-11e8-89f6-2716dbefef31.png)

Reply:
that's very strange, i made it match because when it was being passed into the combine of my function, did a class check to see if I could basically assign. if we had a sentinel scratch state, I could have checked for that,too, and done a no-opp

Reply:
Per discussion, we are going to revert https://github.com/prestodb/presto/commit/3fc4a3e18acab7233033a71ab94bf21bfef416d6 for now . And think about the long-term solution as tracked by https://github.com/prestodb/presto/issues/10186